### PR TITLE
Add more help for "what do do when reading a table fails"

### LINKED ISSUE #6365: Add a "more complicated gallery" to io.ascii output options in the docs
`io.ascii` is a an incredibly complex, but incredibly versatile framework for reading ascii tables. The existing docs focus on describing the existing reader and writer classes, but much more can be done if a user is willing to make a new class. In the discussion of #3812 one such example came up and this PR suggests to add an "example gallery" with a few of those cases.

### PR DESCRIPTION
**Pull request by hamogu**
### Description
Add a gallery of examples to guide users from "oh my table does not read" to "maybe try this option" to eventually writing specific Reader/Writer classes.

Fixes #6365

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will
review this pull request of some common things to look for. This list
is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label.
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] If the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by hamogu** [comment]
@taldcroft I would like your early feedback if you think such a doc addition is useful. I occasionally tell people to do the things I write down here, so I thought it might be useful to add that to the official documentation. I've been writing for a while now, but I'd like your opinion before I go and make sure that all examples work in sphinx etc. (That's why this is still marked as a draft.)
The style is less technical than most of the io.ascii documentation, so you might suggest to put this into astropy learn instead. In that case, I don't want to waste my time getting all the intersphinx right.

On the other hand, my impression is that more people read (or are directed by Google to) the actual astropy documentation, which is why I would like it in there instead of as a "code snippet" on astropy learn or a similar forum.

**Comment by taldcroft** [comment]
I am happy to have the examples live in the astropy docs, and I agree with your reasoning about discovery. There is already precedent in the fixed-width gallery.

**Comment by pllim** [comment]
Maybe we should loop in the Learn team... @kelle @eblur @adrn et al.

**Comment by kelle** [comment]
The Learn team, and @jonathansick in particular, is working on building infrastructure to auto-populate an example gallery from the examples in the Sphinx docs using those `EXAMPLE START` and `EXAMPLE END` tags. So this looks great to us!

File: docs/io/ascii/howtoexamples.rst
Comment (taldcroft) [inline docs/io/ascii/howtoexamples.rst:168]: This seems really contrived and I can't see this being "often" useful.

File: docs/io/ascii/howtoexamples.rst
Comment (taldcroft) [inline docs/io/ascii/howtoexamples.rst:194]: This is not a realistic option for the vast majority of users, having seen some recent work to add a new reader class, it pretty much requires help from the astropy devs. I think the existing https://docs.astropy.org/en/latest/io/ascii/read.html#advanced-customization section gets the point across.

**Review by taldcroft — CHANGES_REQUESTED** [review]
@hamogu - there are some nuggets of useful advice here, but overall I'm suggesting descoping a bit to get this merged.

File: docs/io/ascii/howtoexamples.rst
Comment (hamogu) [reply docs/io/ascii/howtoexamples.rst:5]: That's because I never finished what I indented to write here. This is supposed to get to e.g. https://github.com/astropy/astropy/pull/3812#issuecomment-314771680 which I think *is* useful - there is no guarantee that you and I will be on this project forever, so we should write it down. It can also be a path for others to become developers - that's how I got into astropy (with your help - remember the DAOPhotTableReader?)

File: docs/io/ascii/howtoexamples.rst
Comment (hamogu) [reply docs/io/ascii/howtoexamples.rst:19]: Fixed-width is quite a common format for tables written by legacy FORTRAN code, and sometimes C code, I;ve seen that in PHOENIX, FLASH, PLUTO, and a few others.

Comment (taldcroft) [reply docs/io/ascii/howtoexamples.rst:19]: The question was whether apps like FLASH both (1) emit broken output in one format (e.g. CSV), thus leading the user to this page; and (2) have an option to write data to an alternative format like fixed width or FITS that will not be broken. Generally speaking fixed width output from FORTRAN or C code is going to parse just fine.

**Comment by hamogu** [comment]
I'll leave this on my list and will hopefully work in this and reopen when I have a downscoped version in the next few days.

File: docs/io/ascii/howtoexamples.rst
Comment (hamogu) [reply docs/io/ascii/howtoexamples.rst:140]: This is what I get in IPython or the Jupyter notebook without `guess=False` (copy and paste full output from the terminal. I think that's what most users use these days. We `guess=False`, I would not see "a long list of formats
that it tried on this table"

```
(kitchensink) MoritzAirRoseGold ~> ipython
Python 3.10.1 | packaged by conda-forge | (main, Dec 22 2021, 01:39:07) [Clang 11.1.0 ]
Type 'copyright', 'credits' or 'license' for more information
IPython 7.30.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]:   >>> from astropy.io import ascii
   ...:   >>> table = '''
   ...:   ... Star SpT
   ...:   ... Vega A0
   ...:   ... alpha Cen G2V+K1
   ...:   ... '''
   ...:   >>> ascii.read(table)
---------------------------------------------------------------------------
InconsistentTableError                    Traceback (most recent call last)
~/mambaforge/envs/kitchensink/lib/python3.10/site-packages/astropy/io/ascii/ui.py in _guess(table, read_kwargs, format, fast_reader)
    521             reader = get_reader(**read_kwargs)
--> 522             dat = reader.read(table)
    523             _read_trace.append({'kwargs': copy.deepcopy(read_kwargs),

~/mambaforge/envs/kitchensink/lib/python3.10/site-packages/astropy/io/ascii/core.py in read(self, table)
   1368 
-> 1369                     raise InconsistentTableError(errmsg)
   1370 

InconsistentTableError: Number of header columns (2) inconsistent with data columns (3) at data line 1
Header values: ['Star', 'SpT']
Data values: ['alpha', 'Cen', 'G2V+K1']

During handling of the above exception, another exception occurred:

InconsistentTableError                    Traceback (most recent call last)
<ipython-input-1-732a26b21e0d> in <module>
      5 alpha Cen G2V+K1
      6 '''
----> 7 ascii.read(table)

~/mambaforge/envs/kitchensink/lib/python3.10/site-packages/astropy/io/ascii/ui.py in read(table, guess, **kwargs)
    339         # through below to the non-guess way so that any problems result in a
    340         # more useful traceback.
--> 341         dat = _guess(table, new_kwargs, format, fast_reader)
    342         if dat is None:
    343             guess = False

~/mambaforge/envs/kitchensink/lib/python3.10/site-packages/astropy/io/ascii/ui.py in _guess(table, read_kwargs, format, fast_reader)
    551                    '************************************************************************']
    552             lines.extend(msg)
--> 553             raise core.InconsistentTableError('\n'.join(lines))
    554 
    555 

InconsistentTableError: 
ERROR: Unable to guess table format with the guesses listed below:
Reader:Ecsv fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:FixedWidthTwoLine fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:RST fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:FastBasic fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:Basic fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:FastRdb fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:Rdb fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:FastTab fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:Tab fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:Cds fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:Mrt fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:Daophot fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:SExtractor fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:Ipac fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:Latex fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:AASTex fast_reader: {'enable': True} fill_values: [('', '0')] strict_names: True
Reader:FastCommentedHeader delimiter: '|' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastCommentedHeader delimiter: '|' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastCommentedHeader delimiter: ',' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastCommentedHeader delimiter: ',' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastCommentedHeader delimiter: ' ' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastCommentedHeader delimiter: ' ' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastCommentedHeader delimiter: '\\s' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastCommentedHeader delimiter: '\\s' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:CommentedHeader delimiter: '|' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:CommentedHeader delimiter: '|' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:CommentedHeader delimiter: ',' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:CommentedHeader delimiter: ',' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:CommentedHeader delimiter: ' ' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:CommentedHeader delimiter: ' ' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:CommentedHeader delimiter: '\\s' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:CommentedHeader delimiter: '\\s' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastBasic delimiter: '|' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastBasic delimiter: '|' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastBasic delimiter: ',' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastBasic delimiter: ',' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastBasic delimiter: ' ' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastBasic delimiter: ' ' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastBasic delimiter: '\\s' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastBasic delimiter: '\\s' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:Basic delimiter: '|' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:Basic delimiter: '|' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:Basic delimiter: ',' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:Basic delimiter: ',' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:Basic delimiter: ' ' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:Basic delimiter: ' ' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:Basic delimiter: '\\s' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:Basic delimiter: '\\s' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastNoHeader delimiter: '|' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastNoHeader delimiter: '|' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastNoHeader delimiter: ',' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastNoHeader delimiter: ',' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastNoHeader delimiter: ' ' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastNoHeader delimiter: ' ' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:FastNoHeader delimiter: '\\s' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:FastNoHeader delimiter: '\\s' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:NoHeader delimiter: '|' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:NoHeader delimiter: '|' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:NoHeader delimiter: ',' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:NoHeader delimiter: ',' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:NoHeader delimiter: ' ' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:NoHeader delimiter: ' ' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:NoHeader delimiter: '\\s' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: '"' strict_names: True
Reader:NoHeader delimiter: '\\s' fast_reader: {'enable': True} fill_values: [('', '0')] quotechar: "'" strict_names: True
Reader:Basic fast_reader: {'enable': True} fill_values: [('', '0')]

************************************************************************
** ERROR: Unable to guess table format with the guesses listed above. **
**                                                                    **
** To figure out why the table did not read, use guess=False and      **
** fast_reader=False, along with any appropriate arguments to read(). **
** In particular specify the format and any known attributes like the **
** delimiter.                                                         **
************************************************************************

```

**Comment by pllim** [comment]
This still looks stale, so I am remilestoning.

**Comment by hamogu** [comment]
@taldcroft I pushed changes that hopefully address your comments a while ago. I guess you were waiting for me to fix the remaining sphinx issues and I was waiting for you to broadly say "looks like the right direction" before I dot the i's and cross the t's. 
The previous logs are not available any longer, so I rebased and pushed again, but please let me know how you want to proceed - should I make sure all tests come back green before you look at it or do you have time to look first so I don't waste time fixing um formatting in sections that you later ask me to remove anyway?

**Comment by taldcroft** [comment]
@hamogu - this just fell off my radar. I'll have a look.

**Comment by taldcroft** [comment]
> @taldcroft I pushed changes that hopefully address your comments a while ago.

@hamogu - I don't see any sign of changes since my comments. Normally GitHub would be showing `Outdated` where changes have been made and I would see applicable commits. Looking again at the current file diffs, I can verify that none of the comments I made have been addressed (ranging from big picture stuff like delete this whole section down to fixing typos).

We might want to get review from someone else because I think we have pretty different visions on this and maybe mine is not the best. For instance @mhvk gave this a :+1: and @dhomeier is a maintainer and @bsipocz has commented on `io.ascii` issues.

**Comment by hamogu** [comment]
Mmm. I thought I had addressed them and GH did show me an "outdated", but then I rebased and force pushed. So, I probably forces-pushed from a branch that did not have the newest version and overwrite the changes that I had already made. I'll fix that and then ping you again.

File: docs/io/ascii/howtoexamples.rst
Comment (hamogu) [reply docs/io/ascii/howtoexamples.rst:12]: When I encounter a table that does not read, trying `guess=False` is usually not the first thing I do. I usually try to get a different format (fits, csv instead of space-separated etc.) first because then I don't have ot find out what the options are that allow me to read it with the ascii readers.

**Comment by hamogu** [comment]
@taldcroft Please look again. I see a lot of "outdated" now. I will split off the content from https://github.com/astropy/astropy/pull/3812#issuecomment-314797798 in the separate PR since that's a lot more advanced and we might agree on having this PR, but not the other one.

File: docs/io/ascii/read.rst
Comment (taldcroft) [inline docs/io/ascii/read.rst:873]:
```suggestion
Obtain the Data Table in a Different Format
```

File: docs/io/ascii/read.rst
Comment (taldcroft) [inline docs/io/ascii/read.rst:895]: (See https://docs.astropy.org/en/latest/development/style-guide.html#abbreviations)
```suggestion
Sometimes, the parameters for `astropy.io.ascii.read` to specify, for example
```

File: docs/io/ascii/read.rst
Comment (taldcroft) [inline docs/io/ascii/read.rst:907]:
```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: docs/io/ascii/read.rst
Comment (taldcroft) [inline docs/io/ascii/read.rst:955]:
```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: docs/io/ascii/read.rst
Comment (taldcroft) [inline docs/io/ascii/read.rst:984]:
```suggestion
around ``"alpha Cen"``. Then we can try to read the table again and see
```

File: docs/io/ascii/read.rst
Comment (taldcroft) [inline docs/io/ascii/read.rst:868]:
```suggestion
How to Find and Fix Problems Reading a Table
============================================
```

**Review by taldcroft — CHANGES_REQUESTED** [review]
Thanks @hamogu - basically this looks good now, I just have a bunch of formatting / style suggestions. There is one substantial suggestion to expand on finding the problem with an example and explaining things a bit more.

**Comment by hamogu** [comment]
I agree with all suggested changes and merged them in. Thanks for reviewing! Now, I'll just need to check if there are any formatting problems that the CI flags.

**Comment by pllim** [comment]
Just some trailing whitespace. 😉 

```
trim trailing whitespace.................................................Failed
- hook id: trailing-whitespace
- duration: 0.18s
- exit code: 1
- files were modified by this hook
```

File: docs/io/ascii/read.rst
Comment (pllim) [inline docs/io/ascii/read.rst:912]: You might need ellipses (sp?) directive for the doctest, if this is doctested, but let's wait for the CI to complain first after you fix the trailing whitespace.

p.s. You can set your IDE to auto-remove trailing whitespace, I think.

**Review by taldcroft — APPROVED** [review]
Thanks @hamogu !!

**Pull request merged**
